### PR TITLE
Add start and finish date to UpdateListOptions

### PR DESCRIPTION
**Pull request by varoOP**
Please see: https://myanimelist.net/forum/?topicid=2083936

### CODE REVIEW

**Comment by varoOP** [comment]
Let me know if any changes are needed in this PR and I will make them!

@nstratos

File: mal/useranimelist.go [hunk 1]
```diff
@@ -196,6 +198,32 @@ type Comments string
 func (c Comments) updateMyAnimeListStatusApply(v *url.Values) { v.Set("comments", string(c)) }
 func (c Comments) updateMyMangaListStatusApply(v *url.Values) { v.Set("comments", string(c)) }
 
+// StartDate is an option that allows to update the start date of anime and manga
+// in the user's list.
+type StartDate time.Time
+
+// YYYYMMDD is the layout string for the start date and finish date of anime
+// and manga in the user's list.
+const YYYYMMDD = "2006-01-02"
```
Comment (nstratos) [inline mal/useranimelist.go:207]: We need to also be able to remove the values of start and finish dates.
A good way to achieve that is to use the zero value of `time.Time` as we cannot pass `nil`.

Example:

```go
s, _, err := c.Anime.UpdateMyListStatus(ctx, 967,
	mal.AnimeStatusWatching,
	mal.StartDate(time.Date(2022, 02, 20, 0, 0, 0, 0, time.UTC)),
	mal.FinishDate(time.Time{}), // Remove the value of finish date
)
```
Therefore, I am suggest to replace the constant with a small function that handles the zero value.
```suggestion
func formatMALDate(d time.Time) string {
	if d.IsZero() {
		return ""
	}
	return d.Format("2006-01-02")
}
```

File: mal/useranimelist.go [hunk 2]
```diff
@@ -196,6 +198,32 @@ type Comments string
...
+
+func (d StartDate) updateMyAnimeListStatusApply(v *url.Values) {
+	v.Set("start_date", time.Time(d).Format(YYYYMMDD))
```
Comment (nstratos) [inline mal/useranimelist.go:210]:
```suggestion
	v.Set("start_date", formatMALDate(time.Time(d)))
```

File: mal/useranimelist_test.go [hunk 3]
```diff
@@ -134,6 +136,8 @@ func TestAnimeServiceUpdateMyListStatus(t *testing.T) {
 		RewatchValue(1),
 		Tags{"foo", "bar"},
 		Comments("comments"),
+		StartDate(time.Date(2022, 02, 20, 0, 0, 0, 0, time.UTC)),
+		FinishDate(time.Date(2022, 04, 16, 0, 0, 0, 0, time.UTC)),
```
Comment (nstratos) [inline mal/useranimelist_test.go:140]: We can cover adding and removing dates in the same test e.g. add a start date and remove a finish date.
```suggestion
		FinishDate(time.Time{}),
```

File: mal/useranimelist_test.go [hunk 4]
```diff
@@ -106,14 +106,16 @@ func TestAnimeServiceUpdateMyListStatus(t *testing.T) {
 	mux.HandleFunc("/anime/1/my_list_status", func(w http.ResponseWriter, r *http.Request) {
 		testMethod(t, r, http.MethodPatch)
 		testContentType(t, r, "application/x-www-form-urlencoded")
-		testBody(t, r, "comments=comments&is_rewatching=true&num_times_rewatched=2&num_watched_episodes=3&priority=2&rewatch_value=1&score=8&status=completed&tags=foo%2Cbar")
+		testBody(t, r, "comments=comments&finish_date=2022-04-16&is_rewatching=true&num_times_rewatched=2&num_watched_episodes=3&priority=2&rewatch_value=1&score=8&start_date=2022-02-20&status=completed&tags=foo%2Cbar")
```
Comment (nstratos) [inline mal/useranimelist_test.go:109]:
```suggestion
		testBody(t, r, "comments=comments&finish_date=&is_rewatching=true&num_times_rewatched=2&num_watched_episodes=3&priority=2&rewatch_value=1&score=8&start_date=2022-02-20&status=completed&tags=foo%2Cbar")
```

File: mal/useranimelist_test.go [hunk 5]
```diff
@@ -106,14 +106,16 @@ func TestAnimeServiceUpdateMyListStatus(t *testing.T) {
...
 		const out = `
 		{
 		  "status": "completed",
 		  "score": 8,
 		  "num_episodes_watched": 3,
 		  "is_rewatching": true,
 		  "updated_at": "2018-04-25T15:59:52Z",
+		  "start_date": "2022-02-20",
+		  "finish_date": "2022-04-16",
```
Comment (nstratos) [inline mal/useranimelist_test.go:118]:
```suggestion
```

File: mal/useranimelist_test.go [hunk 6]
```diff
@@ -150,6 +154,8 @@ func TestAnimeServiceUpdateMyListStatus(t *testing.T) {
 		Tags:               []string{"foo", "bar"},
 		Comments:           "comments",
 		UpdatedAt:          time.Date(2018, 04, 25, 15, 59, 52, 0, time.UTC),
+		StartDate:          "2022-02-20",
+		FinishDate:         "2022-04-16",
```
Comment (nstratos) [inline mal/useranimelist_test.go:158]: We could leave this empty to show explicitly that we are expecting the value to be empty or we could even remove the line entirely.
```suggestion
		FinishDate:         "",
```

File: mal/useranimelist.go [hunk 7]
```diff
@@ -83,6 +83,8 @@ type AnimeListStatus struct {
 	RewatchValue       int         `json:"rewatch_value"`
 	Tags               []string    `json:"tags"`
 	Comments           string      `json:"comments"`
+	StartDate          string      `json:"start_date,omitempty"`
+	FinishDate         string      `json:"finish_date,omitempty"`
```
Comment (nstratos) [inline mal/useranimelist.go:87]: This struct is only used to unmarshal from the MAL API so there's no need for omitempty here.
```suggestion
	StartDate          string      `json:"start_date"`
	FinishDate         string      `json:"finish_date"`
```

**Review by nstratos — COMMENTED** [review]
Hey @varoOP 

I added some suggestions that will allow to also remove the dates.

Thank you very much for adding this feature! 

Excellent work! 🚀

**Pull request merged**